java/core/src/main/java/com/google/protobuf/Utf8.java

```diff
@@ -613,8 +614,34 @@ private static int partialIsValidUtf8(final ByteBuffer buffer, int index, final
      *
      * @throws InvalidProtocolBufferException if the byte array slice is not valid UTF-8.
      */
-    abstract String decodeUtf8(byte[] bytes, int index, int size)
-        throws InvalidProtocolBufferException;
+    String decodeUtf8(byte[] bytes, int index, int size)
+        throws InvalidProtocolBufferException {
+      try {
+        String s = new String(bytes, index, size, Internal.UTF_8);
+
+        // "\uFFFD" is UTF-8 default replacement string, which illegal byte sequences get replaced with.
+        if (!s.contains("\uFFFD")) {
+          return s;
+        }
+
+        // Since s contains "\uFFFD" there are 2 options:
+        // 1) The byte array slice is invalid UTF-8.
+        // 2) The byte array slice is valid UTF-8 and contains encodings for "\uFFFD".
+        // To rule out (1), we encode s and compare it to the byte array slice.
+        // If the byte array slice was invalid UTF-8, then we would get a different sequence of bytes.
+
+        // TODO(amirhadadi): When targeting Java 9, replace with Arrays.equals that accepts from and to indexes
+        // to avoid copying the byte array slice.
+        if (Arrays.equals(s.getBytes(Internal.UTF_8), Arrays.copyOfRange(bytes, index, index + size))) {
+          return s;
+        }
+
+        throw InvalidProtocolBufferException.invalidUtf8();
+      } catch (final IndexOutOfBoundsException e) {
+        throw new ArrayIndexOutOfBoundsException(
```

Comment:
This should be an InvalidProtocolBufferException here, not a runtime exception.

Reply:
The catch clause is catching the `StringIndexOutOfBoundsException` thrown from `String` constructor.
`InvalidProtocolBufferException` is not caught.
I think I'm missing your point.

Reply:
I'm preserving the [current behavior](https://github.com/protocolbuffers/protobuf/blob/d8ccfbf00541eb0705bbba640d65e3df814b3d2a/java/core/src/main/java/com/google/protobuf/Utf8.java#L1370) of throwing `ArrayIndexOutOfBoundsException` in case the bytes slice is not within the bounds of the array.

Reply:
@elharo anything else should be done here?